### PR TITLE
Tolerate Escaped Right Braces And Empty Groups

### LINKED ISSUE #375: Hide snippet (or $0) in suggestions
Hello,

Is there a way to hide the snippet, or the $0 in suggestions list ? It's not really nice to see the $0 at the end of every line.

E.G
screenshot

Thank you for your answer,
Peekmo

### PR DESCRIPTION
**Pull request by joefitzgerald**
- Fixes #375 

https://github.com/atom/snippets/pull/123 allows snippets to contain escaped right braces (`\\}`). This might be useful - for example - to display `interface{}` within a snippet's body (as is common for `go` suggestions). This PR allows them to be displayed correctly in the suggestion list. Additionally, it resolves #375, by removing empty snippet groups from the text displayed in the suggestion list.

E.g.:

* `hello$0` > `hello`
* `hello${0}` > `hello`
* `hello${0:}` > `hello`
* `${0:hello}` > `<span class="snippet-completion">hello</span>`
* `${0:hello{\\}}` > `<span class="snippet-completion">hello{}</span>`

When reviewing the code, you may find the lack of regex ugly, or overly complex. However, without the presence of lookbehind in javascript regexes, an alternative approach is required to differentiate an escaped right brace from an unescaped right brace.

Additionally, a small epilogue:

This also does away with snippet markers; they seem unnecessary. Implicitly, the moment a person who is typing reaches the first snippet, their replacement prefix will no longer match - the snippetMarker (`|`) will not match the text in almost all scenarios.

### CODE REVIEW

File: lib/suggestion-list-element.coffee [hunk 1]
```diff
@@ -188,24 +186,17 @@ class SuggestionListElement extends HTMLElement
     else
       rightLabelSpan.textContent = ''
 
-  getHighlightedHTML: (text, snippet, replacementPrefix) ->
-    # 1. Pull the snippets out, replacing with placeholder
-    # 2. Highlight relevant characters
-    # 3. Place snippet HTML back at the placeholders
+  getHighlightedHTML: (text, snippet, replacementPrefix) =>
```
Comment (benogle) [inline lib/suggestion-list-element.coffee]: Why the fat arrow?

File: lib/suggestion-list-element.coffee [hunk 2]
```diff
@@ -222,16 +213,105 @@ class SuggestionListElement extends HTMLElement
...
     # ->   highlightedHTML: '<span class="character-match">a</span>bc(<span class="snippet-completion">d</span>, <span class="snippet-completion">e</span>)f'
-    if snippetCompletions.length
-      completionIndex = 0
-      highlightedHTML = highlightedHTML.replace @snippetMarkerRegex, (match, snippetText) ->
-        snippetCompletions[completionIndex++]
-
+    highlightedHTML = @enhanceSnippet(highlightedHTML) if _.isString(snippet)
     highlightedHTML
 
+  enhanceSnippet: (text) =>
```
Comment (benogle) [inline lib/suggestion-list-element.coffee]: Can we name this `wrapSnippetsWithHTML` or something similar? 

Why the fat arrow?

**Comment by benogle** [comment]
> This also does away with snippet markers; they seem unnecessary. Implicitly, the moment a person who is typing reaches the first snippet, their replacement prefix will no longer match - the snippetMarker (|) will not match the text in almost all scenarios.

The markers are there so the snippets dont get highlighted in the char highlight process. There are tests for this, but you've edited them so they no longer test what they're supposed to. See https://github.com/atom-community/autocomplete-plus/issues/301 and https://github.com/atom-community/autocomplete-plus/commit/37f21515013fb3e47d79e5827add2a89155f00f8

File: spec/autocomplete-manager-integration-spec.coffee [hunk 3]
```diff
@@ -336,7 +335,7 @@ describe 'Autocomplete Manager', ->
               charMatch = editorView.querySelector('.autocomplete-plus li span.word .character-match')
               expect(word.textContent).toBe 'ab(c)c'
               expect(charMatch.textContent).toBe 'c'
-              expect(charMatch.parentNode).toHaveClass 'word'
+              expect(charMatch.parentNode.parentNode).toHaveClass 'word'
```
Comment (benogle) [inline spec/autocomplete-manager-integration-spec.coffee]: This makes the test invalid. The entire point of this test is this line. In this new code, the first `c`, the snippet, will be highlighted, when the second `c` is the one that should be highlighted.

File: lib/suggestion-list-element.coffee [hunk 4]
```diff
@@ -222,16 +213,105 @@ class SuggestionListElement extends HTMLElement
...
-    # e.g. highlightedHTML: '<span class="character-match">a</span>bc(|, |)f'
+    # Remove / replace snippet metadata to make it more readable
+    # e.g. highlightedHTML: '<span class="character-match">a</span>bc(${d}, ${e})f'
     # ->   highlightedHTML: '<span class="character-match">a</span>bc(<span class="snippet-completion">d</span>, <span class="snippet-completion">e</span>)f'
-    if snippetCompletions.length
-      completionIndex = 0
-      highlightedHTML = highlightedHTML.replace @snippetMarkerRegex, (match, snippetText) ->
-        snippetCompletions[completionIndex++]
-
+    highlightedHTML = @enhanceSnippet(highlightedHTML) if _.isString(snippet)
```
Comment (benogle) [inline lib/suggestion-list-element.coffee]: Can we find another way to not pass all of the html into this method? What if there is something in the HTML that matches your snippet state machine?

**Comment by benogle** [comment]
I'm fine with the non-regex parser, but i think we need to come up with a way to separate the html generation of the snippets with that of the char highlights. My thinking:

1. Run the snippet parser. This will give you indices into where they are, and you can gen the HTML for them.
1. Run the char match highlighting code _without_ the snippets in there. My solution was the markers, which maybe arent ideal.
1. Compose the char highlighted text with the snippet highlighted text. 
1. Return the composed text

File: lib/suggestion-list-element.coffee [hunk 5]
```diff
@@ -222,16 +213,105 @@ class SuggestionListElement extends HTMLElement
...
+          snippetStartIndex = -1
+          snippetEndIndex = -1
+          bodyStartIndex = -1
+          bodyEndIndex = -1
+
+      unless (inSnippet or inSnippetBody)
+        text += char
+        continue
+
+    return text
```
Comment (benogle) [inline lib/suggestion-list-element.coffee]: just `text`

File: lib/suggestion-list-element.coffee (hunk 2 above)
Comment (joefitzgerald) [inline lib/suggestion-list-element.coffee]: Fixed in 4fda987dbf3fc7dfa091859c31e4bf97beade625.

File: lib/suggestion-list-element.coffee (hunk 1 above)
Comment (joefitzgerald) [inline lib/suggestion-list-element.coffee]: Fixed in 4fda987dbf3fc7dfa091859c31e4bf97beade625.

File: lib/suggestion-list-element.coffee (hunk 5 above)
Comment (joefitzgerald) [inline lib/suggestion-list-element.coffee]: Fixed in 4fda987dbf3fc7dfa091859c31e4bf97beade625.

File: spec/autocomplete-manager-integration-spec.coffee (hunk 3 above)
Comment (joefitzgerald) [inline spec/autocomplete-manager-integration-spec.coffee]: @benogle I think the tests you have are invalid. It should be valid to have a character match for something within the snippet body. It's just not valid to match any of the HTML wrapping a snippet (i.e. `<span class="snippet-completion">` or `</span>`) or the snippet definition (i.e. `${0:` or `}`).

With my new logic, I'm able to stably determine the snippet definition text and exclude it from character matches, prior to insertion of snippet HTML. Consequently, the tests need to change. I think the intent of #301 is to ensure that:

1. Snippet HTML does not make its way into the suggestion list
1. Snippet HTML is not highlighted
1. Snippet definition text does not make its way into the suggestion list
1. Snippet definition text does not cause incorrect highlighting of character matches

This is now achieved with my latest changes, that I am about to push.

**Comment by joefitzgerald** [comment]
@benogle :eyes:

**Comment by benogle** [comment]
Looking...

File: lib/suggestion-list-element.coffee [hunk 6]
```diff
@@ -226,49 +224,172 @@ class SuggestionListElement extends HTMLElement
...
-        snippetCompletions[completionIndex++]
+      highlightedHTML += '<span class="character-match">' + replacement[i] + '</span>'
 
     highlightedHTML
 
+  removeEmptySnippets: (text) ->
+    return text unless text?.length and text.indexOf('$') isnt -1 # No snippets
+    text.replace(@emptySnippetGroupRegex, '') # Remove all occurrences of $0 or ${0} or ${0:}
+
+  removeAllSnippets: (text) ->
```
Comment (benogle) [inline lib/suggestion-list-element.coffee]: Looks like nothing calls this method?

Comment (joefitzgerald) [inline lib/suggestion-list-element.coffee]: Yes, need to remove that. Ignore that function.

File: spec/autocomplete-manager-integration-spec.coffee (hunk 3 above)
Comment (benogle) [inline spec/autocomplete-manager-integration-spec.coffee]: If you are going to change the spec, check that the parent node is the snippet span class, not this parent.parent indirect check.

**Comment by benogle** [comment]
Overall, it uses a lot of arrays, and this function will be run _a lot_. Like several times for each list display. I feel like the `skipChar` arrays can be removed. The approach in my head is similar to yours, just with  different `replacement` var contents:

* Run `findSnippets` to find the `start`, `end`, `snippetTextStart`, `snippetTextEnd`
* Assemble the text as it would be displayed to the user (call it `replacementText`)
* Run that text through `findCharacterMatches` to get a list of indices
* Use the same algorithm you have now to walk `replacementText` and assemble the HTML based on the indices you have, using math on snippetTextStart

e.g. 

```coffee
snippet = 'text(${1:ab}, ${1:cd})'

# Run findSnippets(snippet)
snippets = [{
  start: 5
  end: 11
  textStart: 9
  textEnd: 10
  text: 'ab'
}, {
  start: 14
  end: 20
  textStart: 18
  textEnd: 19
  text: 'ab'
}]

# Use snippets to assemble this
replacementText = 'text(ab, cd)'

# Mung symbols to have indices into `replacementText`
snippets = [{
  start: 5
  end: 11
  textStart: 9
  textEnd: 10
  replacementStart: 5
  replacementEnd: 6
  text: 'ab'
}, {
  start: 14
  end: 20
  textStart: 18
  textEnd: 19
  replacementStart: 9
  replacementEnd: 10
  text: 'ab'
}]

# Run findCharacterMatches(replacementText)
characterMatches = [0, 2, 5]

displayHTML = ''
for replacementChar, index in replacementText
  # update displayHTML using your indices.
  
return displayHTML
```

**Comment by joefitzgerald** [comment]
Got any idea why the symbol-provider-specs are failing?

**Comment by benogle** [comment]
It's a flaky test. It almost always fails on the `pr` build.

**Comment by benogle** [comment]
Also, you could do away with the indexOf checks with 2 objects: one for snippet starts/ends, and one for char matches. So maybe `findCharacterMatches` returns an object `{0: true, 2: true, 5: true}`, and instead of adding `replacementStart/End` to the snippets array, you create another object: `{5: 'start', 6: 'end', ...}`

**Comment by joefitzgerald** [comment]
I'm not sure switching the snippet info over to 'start' / 'end' saves anything;  is it just to avoid indexOf checks? The storage of the information will be larger.

**Comment by benogle** [comment]
It saves allll the skipChar arrays.

**Comment by benogle** [comment]
Wait, you're referring to the last comment. Yes, it's saving the instance of checks, or iterating the snippets array. 

Also, I'm not optimizing for absolute storage as this stuff is temp. I want to optimize for minimal garbage and cycles.

**Comment by benogle** [comment]
> I'm not sure switching the snippet info over to 'start' / 'end' saves anything; is it just to avoid indexOf checks? The storage of the information will be larger.

And to be clear, I meant creating an object with the `'start'` and `'end'` in place of the `Mung snippets` step.

**Comment by joefitzgerald** [comment]
@benogle :eyes: on the latest changes please.

**Comment by benogle** [comment]
Ok looking / tinkering

**Comment by joefitzgerald** [comment]
Those tests seem to fail pretty reliably for stable reasons. Could it be they need modifying?

**Comment by benogle** [comment]
> Could it be they need modifying?

Yes, they should be fixed to reliably pass. 

I've left feedback in the form of a PR: https://github.com/atom-community/autocomplete-plus/pull/396

**Pull request closed without merging**